### PR TITLE
New dependency resolving algorithm, better error handling, comprehensive tests

### PR DESCRIPTION
**Pull request by amwolff**
The patch introduces following changes:
1. Replaced old dependency resolving solution with depth-first-search-based topological sorting algorithm. This also addresses #8.
2. Revised and changed error handling in the package and in the sub-packages keeping in mind that this code shouldn't ever panic.
3. Wrote tests covering non-trivial code.

### CODE REVIEW

File: jobhub.go [hunk 1]
```diff
@@ -72,47 +66,40 @@ var statusDescription = map[StatusCode]string{
...
+		name:          name,
+		jobByID:       make(map[int]Job),
+		jobDependency: make(map[int][]int),
 	}
 }
 
 func nextIDPipeline() int {
-	tempID := nextPipelineID
-	nextPipelineID++
+	tempID := pipelineIDPool
```
Comment (orian) [inline jobhub.go]: pool usually means that we have a limited number of some resource, like dababase connection pool or file descriptor pool, which needs to be shared across multiple users.

File: jobhub.go [hunk 2]
```diff
@@ -72,47 +66,40 @@ var statusDescription = map[StatusCode]string{
...
-}
-
-func NewPipeline() *Pipeline {
-	return &Pipeline{
-		Log:             logrus.StandardLogger(),
-		jobByID:         make(map[int]Job),
-		jobDependency:   make(map[int][]int),
-		startingJob:     make(map[int]bool),
-		recursionLevels: make(map[int]int),
+func NewPipeline(name string) *pipeline {
```
Comment (orian) [inline jobhub.go]: Returning private type is going to be problematic. I'm for a public type.

File: jobhub.go [hunk 3]
```diff
@@ -122,106 +109,95 @@ func (p *Pipeline) AddJob(job Job) Job {
...
-	}
-	p.recursionLevels[jobID] = level
-	level++
-	for _, depID := range p.jobDependency[jobID] {
-		p.resolveDependencyRecursion(depID, level)
-	}
-}
+func (p pipeline) topologicalSort() ([]int, error) {
+	var (
+		temporaryMark = map[int]bool{}
```
Comment (orian) [inline jobhub.go]: What about we use `make` for all new slices we don't plant to have initial values? 
This adds info about the intention "I want an empty map".

File: jobhub.go [hunk 4]
```diff
@@ -122,106 +109,95 @@ func (p *Pipeline) AddJob(job Job) Job {
...
-	var queue []int
-	tempRL := make(map[int]int)
-	for startID, _ := range p.startingJob {
-		p.resolveDependencyRecursion(startID, 1)
-	}
-	for jID, l := range p.recursionLevels {
-		tempRL[jID] = l
+	visit = func(u int) {
+		if permanentMark[u] {
+			return
```
Comment (orian) [inline jobhub.go]: don't need this return

File: jobhub.go [hunk 5]
```diff
@@ -122,106 +109,95 @@ func (p *Pipeline) AddJob(job Job) Job {
...
+		} else if temporaryMark[u] {
+			acyclic = false
+			return
```
Comment (orian) [inline jobhub.go]: ditto

File: jobhub.go [hunk 6]
```diff
@@ -122,106 +109,95 @@ func (p *Pipeline) AddJob(job Job) Job {
...
-	p.recursionLevels[jobID] = level
-	level++
-	for _, depID := range p.jobDependency[jobID] {
-		p.resolveDependencyRecursion(depID, level)
-	}
-}
+func (p pipeline) topologicalSort() ([]int, error) {
+	var (
+		temporaryMark = map[int]bool{}
+		permanentMark = map[int]bool{}
```
Comment (orian) [inline jobhub.go]: Check this out: https://dave.cheney.net/2014/03/25/the-empty-struct

This is just a curiosity, if you use `map[int]struct{}` you eliminate half of the space required by temporary mark.

You can check if element is in map by:
`if _, ok := permanentMark[1]; ok {`

File: jobhub.go (hunk 2 above)
Comment (amwolff) [reply jobhub.go]: Done.

File: jobhub.go (hunk 1 above)
Comment (amwolff) [reply jobhub.go]: Done; renamed to previous.

File: jobhub.go (hunk 3 above)
Comment (amwolff) [reply jobhub.go]: Done.

File: jobhub.go (hunk 4 above)
Comment (amwolff) [reply jobhub.go]: Done.

File: jobhub.go (hunk 5 above)
Comment (amwolff) [reply jobhub.go]: Done.

**Review by orian — APPROVED** [review]

**Pull request merged**